torch/library.py

```diff
@@ -175,7 +176,68 @@ def wrap(f):
     return wrap
 
 
-def define(lib, schema, alias_analysis=""):
+_keep_alive = []
+
+
+@functools.singledispatch
+def define(name, schema, *, lib=None):
+    r"""Defines a new operator.
+
+    In PyTorch, defining an op (short for "operator") is a two step-process:
+    - we need to define the op (by providing an operator name and schema)
+    - we need to implement behavior for how the operator interacts with
+      various PyTorch subsystems, like CPU/CUDA Tensors, Autograd, etc.
+
+    This entrypoint defines the custom operator (the first step)
+    you must then perform the second step by calling various
+    ``impl_*`` APIs.
+
+    Args:
+        name (str): Should be a string that looks like
+            "namespace::operator_name". Operators in PyTorch need a namespace to
+            avoid name collisions; a given operator may only be created once.
+            If you are writing a Python library, we recommend the namespace to
+            be the name of your top-level module.
+        schema (str): The schema of the operator
+        lib (Optional[Library]): If provided, the lifetime of this operator
+            will be tied to the lifetime of the Library object.
+
+    Example::
+        >>> # xdoctest: +REQUIRES(env:TORCH_DOCTEST_LIBRARY)
+        >>> import torch
+        >>> import numpy as np
+        >>>
+        >>> # Define the operator
+        >>> torch.library.define("mylib::sin", "(Tensor x) -> Tensor")
+        >>>
+        >>> # Add implementations for the operator
+        >>> lib = torch.library.Library("mylib", "FRAGMENT")
+        >>> @torch.library.impl(lib, "sin", "CPU")
+        >>> def f(x):
+        >>>     return torch.from_numpy(np.sin(x.numpy()))
+        >>>
+        >>> # Call the new operator from torch.ops.
+        >>> x = torch.randn(3)
+        >>> y = torch.ops.mylib.sin(x)
+        >>> assert torch.allclose(y, x)
+
+    """
+    if not isinstance(name, str):
+        raise ValueError(
+            f"define(name, schema): expected name to be instance of str, "
+            f"got {type(name)}")
+    namespace, name = torch._library.utils.parse_namespace(name)
+    if lib is None:
+        lib = Library(namespace, "FRAGMENT")
+        _keep_alive.append(lib)
+    lib.define(name + schema, alias_analysis="")
+
+
+@define.register
```

Comment:
woah, pretty cool. 
Does this use type annotations underneath?
And if someone passes in a torch.Library as the first argument, does it know to prioritize this old impl because it is annotated and the new API's first arg isn't annotated?

Reply:
functools.single_dispatch switches on the type of the first argument. If the type of the first argument is a Library, we got to the legacy overload. Otherwise, it hits the new function we've created here.